server/storage/mvcc/key_index.go

```diff
@@ -221,11 +221,6 @@ func (ki *keyIndex) compact(lg *zap.Logger, atRev int64, available map[Revision]
 		if revIndex != -1 {
 			g.revs = g.revs[revIndex:]
 		}
-		// remove any tombstone
-		if len(g.revs) == 1 && genIdx != len(ki.generations)-1 {
```

Comment:
Hi @ahrtr I'm still working on this.

I handle this in a different way. However, I still need more time to add more test cases for that.

```diff
diff --git a/server/storage/mvcc/key_index.go b/server/storage/mvcc/key_index.go
index e19068f9c..e8223380e 100644
--- a/server/storage/mvcc/key_index.go
+++ b/server/storage/mvcc/key_index.go
@@ -223,8 +223,10 @@ func (ki *keyIndex) compact(lg *zap.Logger, atRev int64, available map[Revision]
 		}
 		// remove any tombstone
 		if len(g.revs) == 1 && genIdx != len(ki.generations)-1 {
-			delete(available, g.revs[0])
-			genIdx++
+			if g.revs[0].Main < atRev {
+				delete(available, g.revs[0])
+				genIdx++
+			}
 		}
 	}

@@ -243,7 +245,9 @@ func (ki *keyIndex) keep(atRev int64, available map[Revision]struct{}) {
 	if !g.isEmpty() {
 		// remove any tombstone
 		if revIndex == len(g.revs)-1 && genIdx != len(ki.generations)-1 {
-			delete(available, g.revs[revIndex])
+			if g.revs[revIndex].Main < atRev {
+				delete(available, g.revs[revIndex])
+			}
 		}
 	}
 }
```

Reply:
> ```diff
> @@ -223,8 +223,10 @@ func (ki *keyIndex) compact(lg *zap.Logger, atRev int64, available map[Revision]
>  		}
>  		// remove any tombstone
>  		if len(g.revs) == 1 && genIdx != len(ki.generations)-1 {
> -			delete(available, g.revs[0])
> -			genIdx++
> +			if g.revs[0].Main < atRev {
> +				delete(available, g.revs[0])
> +				genIdx++
> +			}
>  		}
>  	}
> ```

It doesn't make sense, because `if g.revs[0].Main < atRev` will be always `false`, so it has the same effect as my change, which is much simpler.

Reply:
Thanks for pointing it out. Yes, when there is only one record and it's not the last generation, we should always keep it.